drivers: rtc: stm32: add calibration feature and several fixes/improvements

faloj opened the pull request:
- Adds set_calibration and get_calibration API functions
- Correctly computes nanosecond value
- Fixes day of the week value
- Fixes potential erroneous values issue due to shadow register being bypassed
- Avoids reset induced time drift
- Remove blocking loops
- Fixes non thread safe operations
- Uses BDC<->bin conversion functions instead of the LL ones
- Sets all the rtc_time fields returned by rtc_stm32_get_time()
- Perform additional checks against arguments received by rtc_stm32_set_time()

This have been tested on an STM32H7

Commit:
drivers: rtc: stm32: add calibration feature

This adds the set_calibration and get_calibration API functions to the
STM32 RTC driver

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

Commit:
drivers: rtc: stm32: fix wrong nanosecond value

SSR register does not contain direct nanosecond value. This adds the
necessary conversion.

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

faloj commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -149,16 +176,87 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
 	timeptr->tm_min = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_MINUTE(rtc_time));
 	timeptr->tm_sec = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_SECOND(rtc_time));
```
tm_wday computation is working but not consistent with values given in reference manual, I hesitated to fix it but it is probably to much nitpicking.

bjarki-trackunit replied:
Could you expand on this? From what I can tell from the [reference manual for stm325858ai](https://www.st.com/resource/en/reference_manual/rm0456-stm32u5-series-armbased-32bit-mcus-stmicroelectronics.pdf) page 2599 the RTC_DR.WDU goes from 1 (monday) to 7 (sunday), while the [struct tm](https://cplusplus.com/reference/ctime/tm/) goes from 0 (monday) to 6 (sunday). so RTC_DR.WDU - 1 seems consistent with at least this reference manual :)

bjarki-trackunit replied:
never mind, it seems you are right, 0 in `struct tm` is sunday, of course, as long as the offset is the same it will work as you pointed out, I say fix it :)

faloj replied:
fixed by [804d899](https://github.com/zephyrproject-rtos/zephyr/pull/62304/commits/804d899d594f98af309a8ab143663df586d0782f)

bjarki-trackunit commented:
Looks good, I would like to resolve the issue you brought up, then I think we can have this PR approved and merged :)

Commit:
drivers: rtc: stm32: fix day of the week conversion

The conversion used was working but not consistent with values given by
reference manuals.

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

Commit:
drivers: rtc: stm32: check if year is supported by RTC

RTC does not support year before 2000 while tm/rtc_time structure
reference year is 1900

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

Commit:
drivers: rtc: stm32: check if tm_wday is set

tm/rtc_time structure allow for tm_wday to be set to -1 if unknown but RTC
is expecting it

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

faloj commented:
I added two input checks : one against tm_year and one against tm_wday

faloj commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -156,13 +156,15 @@ static int rtc_stm32_set_time(const struct device *dev, const struct rtc_time *t
+	/* Read subsecond, time and date registers */
+	rtc_subsecond = LL_RTC_TIME_GetSubSecond(RTC);
+	rtc_time      = LL_RTC_TIME_Get(RTC);
+	rtc_date      = LL_RTC_DATE_Get(RTC);
```
Reference manual says : "To ensure consistency between the 3 values, reading either RTC_SSR or RTC_TR locks the values in the higher-order calendar shadow registers until RTC_DR is read".
As I understand it reading RTC_SSR locks RTC_TR and RTC_DR thus allowing consistent reading. So there is no need to do the "do {} while();" thing that is done in #61580 imho if reading is done in the correct order.

gautierg-st replied:
I didn't know that, I just remembered what was done in the other PR.
So I looked a bit more into it, and that's only valid if you use the shadow registers. But STM32F2x series doesn't have them (and F1x but that's a completely different RTC), and in the current code, LL_RTC_EnableShadowRegBypass is called in the init function anyway. So in any case, the shadow registers are not used, and the "do {} while();" is necessary to ensure consistency.
Even if you remove the call and actually use the shadow register, you will still need a special read for F2x.

faloj replied:
Oh waw ! Indeed Shadow register bypass is enabled… Don't know why… Ok so I will push a fix.

faloj replied:
It may have been enabled so RTC works correctly for any clock configuration. (One must ensure that APB clock is more than seven times the RTCCLK frequency to use shadow registers).
Requested fix is done in [01ce509](https://github.com/zephyrproject-rtos/zephyr/pull/62304/commits/01ce5094652f0be7acf4cd188779564ab76d3826).

faloj commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -141,7 +141,16 @@ static int rtc_stm32_set_time(const struct device *dev, const struct rtc_time *t
+		/* tm_wday == 0 means sunday in rtc_time structure while according to reference
+		 * manual it is represented by a 7 in hardware (and 0 is a forbidden value)
+		 */
+		LL_RTC_DATE_SetWeekDay(RTC, 7);
```
You are right. I fixed it.

Commit:
drivers: rtc: stm32: avoid reset induced time drift

Calling LL_RTC_Init causes the RTC to stop while being configured thus
inducing time drift. This commit avoids doing it at each reset if the RTC
is already properly configured.

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

Commit:
drivers: rtc: stm32: avoid potential erroneous readings

As shadow registers bypass is enabled, an erroneous reading may occur at
each day or second increment. This commit fixes this issue.

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

Commit:
drivers: rtc: stm32: avoid infinite blocking loop

Use the less CPU intensive and timeout capable WAIT_FOR zephyr provided
macro

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

Commit:
drivers: rtc: stm32: avoid using LL_RTC_Init

LL_RTC_Init() internally uses LL_RTC_EnterInitMode() wich uses a blocking
while loop. This rewrite uses the less CPU time wasteful local
rtc_stm32_enter_initialization_mode() instead. As a side effect
LL_RTC_InitTypeDef in rtc_stm32_config is not needed anymore.

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

Commit:
drivers: rtc: stm32: ensure thread safe operations

This adds mutex a lock in rtc_stm32_set_time and rtc_stm32_get_time to
ensure consistent data reading. Also performs register reading in a single
operation in rtc_stm32_get_calibration for the same reason.

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

Commit:
drivers: rtc: stm32: use Zephyr BCD<->bin conversion functions

This replaces the verbose LL_RTC_CONVERT_BCD2BIN and LL_RTC_CONVERT_BIN2BCD
by the bcd2bin and bin2bcd Zephyr provided equivalent

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

Commit:
drivers: rtc: stm32: fill all rtc_time fields

Set tm_yday and tm_isdst to -1 as theses values are unknown.

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

faloj commented:
I have some rework and improvement and updated the first comment consequently.

erwango commented on drivers/rtc/rtc_ll_stm32.c:
```diff
@@ -323,8 +345,10 @@ static int rtc_stm32_get_calibration(const struct device *dev, int32_t *calibrat
 
-	uint32_t calp_enabled = LL_RTC_CAL_IsPulseInserted(RTC);
-	uint32_t calm = LL_RTC_CAL_GetMinus(RTC);
+	uint32_t calr = sys_read32(&RTC->CALR);
+
+	bool calp_enabled = READ_BIT(calr, RTC_CALR_CALP);
+	uint32_t calm = READ_BIT(calr, RTC_CALR_CALM);
```
Unfortunately this isn't compatible with stm32f2 series (which uses different register names).

faloj replied:
Fixed by [1a95dfc](https://github.com/zephyrproject-rtos/zephyr/pull/62304/commits/1a95dfc9d16324064af6ed77b895bd362c698d1b). I have disabled calibration for STM32F2 and STM32F2 Cat. 1 as I can't do any code for coarse calibration since I do only work on an H7 right now.

erwango reviewed (changes requested):
One issue detected otherwise looks sane, but @gautierg-st's review required

Commit:
drivers: rtc: stm32: disable calibration for STM32F2 and some STM32L1

STM32F2 and STM32L1 Cat. 1 only provide the coarse calibration feature
which is not supported by the code yet.

Signed-off-by: Johan Lafon <johan.lafon@syslinbit.com>

gautierg-st reviewed (approved):
LGTM. Thanks

bjarki-andreasen reviewed (approved):
Nice work!
